syncer/syncer.go

```diff
@@ -0,0 +1,285 @@
+package syncer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/0xPolygon/polygon-edge/helper/progress"
+	"github.com/0xPolygon/polygon-edge/network/event"
+	"github.com/0xPolygon/polygon-edge/types"
+	"github.com/hashicorp/go-hclog"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+const (
+	syncerLoggerName = "syncer"
+	SyncerProto      = "/syncer/0.2"
+)
+
+var (
+	errTimeout = errors.New("timeout awaiting block from peer")
+)
+
+// XXX: Don't use this syncer for the consensus that may cause fork.
+// This syncer doesn't assume fork. Consensus may be broken.
+// TODO: Add extensibility for fork before merge
+type syncer struct {
+	logger          hclog.Logger
+	blockchain      Blockchain
+	syncProgression Progression
+
+	peerMap         *PeerMap
+	syncPeerService SyncPeerService
+	syncPeerClient  SyncPeerClient
+
+	// Timeout for syncing a block
+	blockTimeout time.Duration
+
+	// Channel to notify WatchSync that a new status arrived
+	newStatusCh chan struct{}
+}
+
+func NewSyncer(
+	logger hclog.Logger,
+	network Network,
+	blockchain Blockchain,
+	blockTimeout time.Duration,
+) Syncer {
+	return &syncer{
+		logger:          logger.Named(syncerLoggerName),
+		blockchain:      blockchain,
+		syncProgression: progress.NewProgressionWrapper(progress.ChainSyncBulk),
+		syncPeerService: NewSyncPeerService(network, blockchain),
+		syncPeerClient:  NewSyncPeerClient(logger, network, blockchain),
+		blockTimeout:    blockTimeout,
+		newStatusCh:     make(chan struct{}),
+		peerMap:         new(PeerMap),
+	}
+}
+
+// Start starts goroutine processes
+func (s *syncer) Start() error {
+	if err := s.syncPeerClient.Start(); err != nil {
+		return err
+	}
+
+	s.syncPeerService.Start()
+
+	go s.initializePeerMap()
+	go s.startPeerConnectionEventProcess()
+
+	return nil
+}
+
+// Close terminates goroutine processes
+func (s *syncer) Close() {
+	close(s.newStatusCh)
+}
+
+// initializePeerMap fetches peer statuses and initializes map
+func (s *syncer) initializePeerMap() {
```

Comment:
I think you can nicely break the responsibility of this method to have more code clarity. The main goal of this method is to be spawned as a background thread (looping on for), in which case the map initialization can be moved to after the `s.syncPeerService.Start()` statement in the syncer's `Start()` method. Instead of `initializePeerMap` this method can be called `listenForPeerEvents` (or something else).

Reply:
Good suggestion
Fixed
8a0176d4d4165062629ea1cd2857ec1fec765a7e
4683c852ede4be183ef6201a009cd7766c3636cc